managed/services/uievents/uievents.go

```diff
@@ -0,0 +1,259 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package uievents provides facility to store UI events.
+package uievents
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/HdrHistogram/hdrhistogram-go"
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
+	"github.com/percona/pmm/managed/services/telemetry"
+	"github.com/sirupsen/logrus"
+	"sort"
+	"time"
+
+	uievents "github.com/percona/pmm/api/uieventspb"
+)
+
+const cleanupAfterHours = 30
```

Comment:
I know that is not ideal and can cause reporting telemetry "from yesterday". But this can happen only if send interval is modified or there is network issue that last 24+h. This should be a very rare case.